[Yong Jun Xi] iP

TobyCyan opened the pull request:
# Mei
> "The shorter way to do many things is to only do one thing at a time." - Mozart ([Source](https://deepstash.com/idea/18151/the-shorter-way-to-do-many-things-is-to-only))

Behold, your **personal task managing assistant**, **Mei**!
With **Mei**, you can:
+ Add
+ Find
+ Delete
+ Mark/ Unmark
+ Save

***ALL*** your tasks!

## Task Types
Here are some of the tasks you may want to save:
1. A basic to-do
2. Deadlines
3. Events

So you may ~~sometimes~~ ALMOST ALWAYS remember to do them on time 🎉 !

## Basic Features
- [x] All the commands listed above!
- [ ] Pretty GUI (Soon 😄)
- [ ] Reminders (Soon 😃)

> [!NOTE]
> You may be wondering, where do the tasks get saved to in your devices?
> Well *fear not*! Here's the `main` method 😉 
```java
public static void main(String[] args) {
    new Mei("./taskdata/tasks.txt").run();
}
```

## Download
Feel free to give Mei a try by download it [here](https://github.com/TobyCyan/ip/releases/tag/Level-9)!!

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

TobyCyan committed:
Add binary file types to .gitignore to ignore these files when pushing.

This is the first commit, for testing our the committing function in sourcetree.

TobyCyan committed:
Complete the level 0: rename, greet, exit increment

The chatbot's name is Mei, and a response manager class is created
to manager all of her responses.

Currently I'm using a hard-coded method of calling divider lines
before and after each response message calls for the sake of showing
 the results, but should find a more automated way to do this.

TobyCyan committed:
Add a response map and dedicated echo function

A response hash map is used to map supported response types to their
responses. This is to ensure that the responses needed can all be
found in the same place, reducing the chances of unintended errors.

An echo function is also added to ensure that all responses we are
trying to give through the response manager are wrapped in the
desired format.

TobyCyan committed:
Complete level-2: add functionality to add and list tasks

Mei is now able to receive input to add task and list out all user's
tasks.

Communication between the Task Manager and Response Manager
is established by storing the singleton instances in the respective
classes.

This is so that the Response Manager can process the user's input to
add or list tasks and prompt the Task Manager to execute the related
methods, then prompting the Response Manager to echo the result back
to the user.

There is space for errors and exceptions. Therefore,
Let's
- Add exception handling for task-related methods.
- Brainstorm more scenarios of the user's input to redirect them to
the appropriate methods and make the correct response.

TobyCyan committed:
Complete level-3: marking and unmarking tasks, add a task class

A new task class that stores its description, status, and other
methods is added to maintain the abstraction of details.

User will be able to mark tasks as done and unmark tasks as undone.
Mei will also respond accordingly to the user's command input.

A concatenation method for string arrays is added for more
customizing of responses.

Optionally, let's add responses when user tries to mark or unmark
an already marked or unmarked task. This is to make the chatbot seem
more intelligent and shows more personality to it.

TobyCyan committed:
Complete level-4: add new task types

TobyCyan committed:
Complete A-TextUiTesting

The program is tested using the input/output redirection technique,
with most commands such as the different types of tasks, listing,
marking, unmarking, and bye.

Furthermore, cases such as listing without any tasks added and
invalid task types are also tested.

Currently the program is still prone to errors due to weak error
handling.

Therefore, let's start working on the next increment which is
handling errors and further test out more cases.

TobyCyan committed:
Add some initial exceptions

TobyCyan committed:
Complete level-6: handle delete tasks and add more tests

TobyCyan committed:
Update constants' naming convention

TobyCyan committed:
Group files into packages and add read/ write task data to file path

TobyCyan committed:
Fix bugs when saving and loading task data

TobyCyan committed:
Move task type check into task manager class

TobyCyan committed:
Add timed task class and fix formatting bugs when read and write

TobyCyan committed:
Trim task description and figure out relative path to store taskdata

TobyCyan committed:
Trim description at constructor instead of toString method

TobyCyan committed:
Add functionality to accept different date/time formats

TobyCyan committed:
Add small changes to readFromFile

TobyCyan committed:
Merge branch 'branch-level-8'

TobyCyan committed:
Add more OOP-like structure for fileaccess and manager packages

TobyCyan committed:
Merge branch 'branch-a-moreoop'

TobyCyan committed:
Merge branch 'add-gradle-support'

TobyCyan committed:
Add gradle and test commands using gradlew

TobyCyan committed:
Merge branch 'branch-add-gradle'

TobyCyan committed:
Add stubs and test classes with empty test methods

TobyCyan committed:
Add more empty test cases of different classes

TobyCyan committed:
Add test cases to run using JUnit

TobyCyan committed:
Merge branch 'branch-add-junit'

TobyCyan committed:
Package the app as a JAR file

TobyCyan committed:
Add new JavaDoc comments on some classes and methods

TobyCyan committed:
Complete JavaDoc documentation for fileaccess package

TobyCyan committed:
Add spacing between the JavaDoc descriptions and annotations

TobyCyan committed:
Complete JavaDoc documentation on the manager package

TobyCyan committed:
Complete JavaDoc documentation for the tasks package and Mei class

TobyCyan committed:
Make non-constants abide to their naming convention

TobyCyan committed:
Add functionality to find tasks based on keyword in the description

TobyCyan committed:
Merge branch 'branch-Level-9'

TobyCyan committed:
Merge branch 'branch-A-CodingStandard'

TobyCyan committed:
Merge branch 'branch-A-JavaDoc'

TobyCyan committed:
Make minor adjustments due to changes from merge conflicts

TobyCyan committed:
Place the main class into its own package

TobyCyan committed:
Place all packages under the main mei package

TobyCyan committed:
Add header comments to all classes

TobyCyan committed:
Install and set up checkstyle plugin

TobyCyan committed:
Merge branch 'branch-A-CheckStyle'

TobyCyan committed:
Import JavaFX GUI into this project

TobyCyan committed:
Add some more styling and greeting user after initialization

TobyCyan committed:
Merge branch 'branch-A-GUI'

TobyCyan committed:
Comment on a few public methods

leeyenshen commented on src/main/java/mei/tasks/Task.java:
```diff
@@ -0,0 +1,76 @@
+     * @param keyword The keyword used to check for a match.
+     * @return true or false depending on whether the description consists of the given keyword.
+     */
+    public boolean isDescriptionContainsKeyword(String keyword) {
```
Maybe to make it sound better, can use doesDescriptionContainKeyword(String keyword)?

leeyenshen commented on src/main/java/mei/tasks/Deadline.java:
```diff
@@ -0,0 +1,45 @@
+     */
+    public Deadline(String description, String deadlineDateTime) {
+        super(description);
+        this.deadlineDateTime = convertDateTimeFormat(deadlineDateTime);
```
Could this potentially return a null value for deadlineDateTime? If no format is found? Perhaps can try having a null exception catch?

leeyenshen commented on src/main/java/mei/manager/TaskManager.java:
```diff
@@ -0,0 +1,239 @@
+        AtomicInteger i = new AtomicInteger();
+
+        return tasks.stream()
+                .filter(task -> task.isDescriptionContainsKeyword(keyword))
```
Good use of AtomicInteger to manage the index inside the stream. This avoids side effects and makes the code more thread-safe

leeyenshen commented on src/main/java/mei/fileaccess/FileRead.java:
```diff
@@ -0,0 +1,80 @@
+        String description = splitFileData[2];
+
+        switch (taskType) {
+        case "ToDo":
```
Good job with the indentation, and making sure it follows the rules!

Kiranlimtl replied:
I like the indentation, it is easy to miss out

leeyenshen commented on src/main/java/mei/manager/InputManager.java:
```diff
@@ -0,0 +1,127 @@
+            try {
+                // First word doesn't fit in any of the commands nor task types.
+                if (!taskManager.isTaskTypeExist(splitInput[0])) {
+                    throw new UnknownUserInputException(ResponseManager.getResponses("UnknownUserInput"));
```
Nice work handling invalid task indices gracefully! The use of exceptions ensures that errors are caught and properly communicated to the user.

leeyenshen commented on src/main/java/mei/tasks/TimedTask.java:
```diff
@@ -0,0 +1,79 @@
+
+public class TimedTask extends Task {
+    /** The possible formats that the user could input their datetime. **/
+    private static final String[] INPUT_FORMATS = new String[] {"d/MM/yyyy HHmm", "d-MM-yyyy HHmm", "yyyy/MM/d HHmm", "yyyy-MM-d HHmm"};
```
Maybe try wrapping this line? As line length should be no longer than 120 chars.

leeyenshen reviewed (commented):
Great work on this implementation! The code is well-structured and adheres to solid OOP principles. I appreciate the clear Javadoc comments, the efficient use of final for immutability, and the well-thought-out switch statement for task handling. Additionally, good job using AtomicInteger for stream indexing and defensive programming for task validation. There are only 1 or 2 coding standards that need changing.

Kiranlimtl commented on src/test/java/manager/ResponseManagerTest.java:
```diff
@@ -0,0 +1,77 @@
+package manager;
```
I think the ResponseManager class has a well-structured response map.

Kiranlimtl commented on src/main/java/mei/exception/DeadlineNotEnoughInfoException.java:
```diff
@@ -0,0 +1,16 @@
+ * The error response is given based on the string array retrieved from the RESPONSE_MAP in the Response Manager
+ * with the key of the same name excluding the exception word at the end.
+ */
+public class DeadlineNotEnoughInfoException extends MeiException {
```
The exceptions are well structured but maybe can consider consolidating similar exceptions like EventNotEnoughInfoException and DeadlineNotEnoughInfoException to reduce redundancy

Kiranlimtl reviewed (commented):
I think overall is very good. Good job

Kiranlimtl reviewed (commented):
I think most of the names are good, I cannot find any naming issues

Kiranlimtl commented:
I think most of the names are good, I cannot find any naming issues

Kiranlimtl commented on src/main/java/mei/manager/InputManager.java:
```diff
@@ -0,0 +1,127 @@
+     * @param taskIndex The task index to handle.
+     * @return true or false depending on whether the task index is problematic or not.
+     */
+    public boolean isTaskIndexProblematic(int taskIndex) {
```
The method name is abit long, maybe can change to isInvalidTaskIndex or validateTaskIndex

Kiranlimtl reviewed (commented):
Overall good i think the naming is really good other than a few

TobyCyan committed:
Add assertions to certain methods

There are several methods in the manager classes that holds certain
assumptions such as the task index should always be valid and task
passed in should never be null.

This ensures that the program can be stopped safely without
accidentally writing invalid tasks into the task data save file.

The user input from the javafx main window should also never be
empty and it is assumed such a way because empty/ invalid inputs
should be handled before-hand.

Let's look out for more assumptions in future extensions and assert
them where required.

TobyCyan committed:
Extract response into classes

There are several methods that are way too long as everything
(i.e. inputs in Input Manager and responses in Response Manager)
are processed and managed in the same methods respectively.

Extracting each response types into their own classes with their
own defined responses makes the code quality better and easier to
manage as any changes to the desired the responses can be made
within the target response class.

Let's,
* continue extracting as many large code blocks as possible into
separate methods.
* make sure the codebase abides SLAP as closely as possible.

TobyCyan committed:
Abstract code blocks in task manager and remove dead code

Abstracting code for processing add task command from the user
for each task types makes the code cleaner to read.

Some unused code from extracting the responses into classes are
also removed to improve code quality and avoid redundant code.

Let's keep up with improving our code quality in future commits.

TobyCyan committed:
Move response package into mei package

TobyCyan committed:
Remove extra comment

TobyCyan committed:
Merge pull request #1 from TobyCyan/branch-A-Assertions

Add assertions to certain methods

TobyCyan committed:
Merge branch 'master' of https://github.com/TobyCyan/ip

TobyCyan committed:
Merge branch 'master' into branch-A-CodeQuality

TobyCyan committed:
Merge pull request #2 from TobyCyan/branch-A-CodeQuality

Extract out new classes and methods for better readability

TobyCyan committed:
Merge branch 'master' of https://github.com/TobyCyan/ip

TobyCyan committed:
Make small changes to the code format for better readability

TobyCyan committed:
Add functionality to undo the latest command

Caching the latest command by reversing it (e.g. add -> delete,
mark -> unmark and vice versa) will allow execution of this
reversed command if undo is given as the input.

A modification to how the tasks are stored as data by including the
add task command is necessary to make the undo more versatile as we
don't have to manually change how is the add task command defined
for each task types in their classes whenever we decide to
change it.

Let's:
* look into adding functionality to undo multiple commands in a row
instead of just the most recent one, maybe maintain a list instead
of a String.
* possibly refactor some parts of the code and extract them out into
a class.

TobyCyan committed:
Reset most recent reversed input after calling undo

TobyCyan committed:
Update comments

TobyCyan committed:
Update the undo function to be able to undo a chain of commands

TobyCyan committed:
Update input manager class header comment

TobyCyan committed:
Update test cases

TobyCyan committed:
Merge pull request #3 from TobyCyan/branch-C-Undo

Add functionality to undo a chain of recent commands

TobyCyan committed:
Merge branch 'master' of https://github.com/TobyCyan/ip

TobyCyan committed:
Rearrange imports in test classes

TobyCyan committed:
Add error handling for invalid date/time formats

TobyCyan committed:
Update method naming convention

TobyCyan committed:
Update code comments

TobyCyan committed:
Add representative image of the user interface

TobyCyan committed:
Update on header comments

TobyCyan committed:
Add images for user guide

TobyCyan committed:
Update user guide

TobyCyan committed:
Update README.md

TobyCyan committed:
Update README.md

TobyCyan committed:
Merge branch 'master' of https://github.com/TobyCyan/ip

TobyCyan committed:
Add new images for user guide

TobyCyan committed:
Update README.md

TobyCyan committed:
Update README.md

TobyCyan committed:
Update README.md

TobyCyan committed:
Move user guide images to docs folder

TobyCyan committed:
Merge branch 'master' of https://github.com/TobyCyan/ip

TobyCyan committed:
Update README.md

TobyCyan committed:
Update README.md

TobyCyan committed:
Update README.md with section links

TobyCyan committed:
Update README.md

TobyCyan committed:
Update README.md with more white spaces

TobyCyan committed:
Merge branch 'master' of https://github.com/TobyCyan/ip

TobyCyan committed:
Change thrown exception type in filewrite

TobyCyan committed:
Fix invalid inputs to mark, unmark and delete

TobyCyan committed:
Fix task-related and date/time bugs

TobyCyan committed:
Fix coding standard violations

TobyCyan committed:
Update user guide
